Sources/AWSSDKSwiftCore/Credential/MetaDataCredentialProvider.swift

```diff
@@ -0,0 +1,303 @@
+
+    init(eventLoop: EventLoop, client: Client, remainingTokenLifetimeForUse: TimeInterval? = nil) {
+        self.metaDataClient = client
+        self.remainingTokenLifetimeForUse = remainingTokenLifetimeForUse ?? 3 * 60
+    }
+    
+    public func getCredential(on eventLoop: EventLoop) -> EventLoopFuture<Credential> {
+        self.lock.lock()
+        let cred = credential
+        self.lock.unlock()
+        
+        if let cred = cred, !cred.isExpiring(within: remainingTokenLifetimeForUse) {
+            // we have credentials and those are still valid
+            return eventLoop.makeSucceededFuture(cred)
+        }
+        
+        // we need to refresh the credentials
+        return self.refreshCredentials(on: eventLoop)
+    }
+    
+    private func refreshCredentials(on eventLoop: EventLoop) -> EventLoopFuture<Credential> {
+        self.lock.lock()
+        defer { self.lock.unlock() }
+        
+        if let future = credentialFuture {
+            // a refresh is already running
+            if future.eventLoop !== eventLoop {
+                // We want to hop back to the event loop we came in case
+                // the refresh is resolved on another EventLoop.
+                return future.hop(to: eventLoop)
+            }
+            return future
+        }
+        
+        credentialFuture = self.metaDataClient.getMetaData(on: eventLoop)
+            .map { (metadata) -> (Credential) in
+                let credential = metadata.credential
+                
+                // update the internal credential locked
+                self.lock.withLock {
+                    self.credentialFuture = nil
+                    self.credential = credential
+                }
+                return credential
+            }
+
+        return credentialFuture!
+    }
+}
+
+struct ECSMetaDataClient: MetaDataClient {
+    public typealias MetaData = ECSMetaData
+    
+    static let Host = "169.254.170.2"
+    static let RelativeURIEnvironmentName = "AWS_CONTAINER_CREDENTIALS_RELATIVE_URI"
+    
+    struct ECSMetaData: CredentialContainer {
+        let accessKeyId: String
+        let secretAccessKey: String
+        let token: String
+        let expiration: Date
+        let roleArn: String
+
+        public var credential: ExpiringCredential {
+            return RotatingCredential(
+                accessKeyId: accessKeyId,
+                secretAccessKey: secretAccessKey,
+                sessionToken: token,
+                expiration: expiration
+            )
+        }
+
+        enum CodingKeys: String, CodingKey {
+            case accessKeyId = "AccessKeyId"
+            case secretAccessKey = "SecretAccessKey"
+            case token = "Token"
+            case expiration = "Expiration"
+            case roleArn = "RoleArn"
+        }
+    }
+    
+    let httpClient    : AWSHTTPClient
+    let endpointURL   : String
+    let decoder       = Self.createJSONDecoder()
```

Comment:
Is JSONDecoder() thread safe?

Reply:
yes, and it doesn't have to be. 😉 will never be called synchronous.